### ISSUE TITLE
World files only in tmpfs/shared memory

### ISSUE THREAD

**Issue by ShadwDrgn** [itzg/docker-minecraft-server#1231]
### Enhancement Type

A completely new feature

### Describe the enhancement

It would be spectacularly nice if i could store the world folder in RAM so that worldgen doesn't cause as much lag. This feature would obviously require the end user to use backups AND would require that the server restore from backup as a part of startup. Not sure how simple this would be to implement using what you've got, but it would be really nice, as the custom container i build now does this and I'd love to switch to your FAR FAR more mature and organized image here, but don't particularly want to go back to world on disk. :D

**Comment by shotah** [comment]
Looking into this as it sounds very interesting. Found a couple articles and I'll need to dig more if this is just something that would need to be done outside of the container and just passed in as the data folder, or something we can control from the docker file or scripts. 

Related Articles:
[Step by Step Guide](https://minecraft.fandom.com/wiki/Tutorials/Ramdisk_enabled_server)
[Reddit Article](https://www.reddit.com/r/admincraft/comments/2sv9x3/ramdisk_for_world_files/)

**Comment by ShadwDrgn** [comment]
I used tmpfs when I did it for mine, but mine requires a rebuild every time i change mods or whitelist or anything because i'm not properly utilizing volumes for stuff like that. also my backup solution is in container instead of a side container (using borg backup which I love). It'd be prefered tto somehow make it a seperate container i can load or not if possible. not sure how that would work alongside having to restore at boot up though. docker doesn't seem to do orchestration of dependency containers or init containers in any reasonable way i've been able to find (kubernetes aside).

**Comment by shotah** [comment]
I guess I am still having a hard time understanding what you are asking for. Docker compose does do container orchestration, and so does kubernetes, but I am not sure what a second container would have to do with the 'docker-minecraft-server' container. Also with setting up tmpfs and backup jobs those again are all defined out side of the 'docker-minecraft-server' container. So what do  you see being implemented inside of the container?

**Comment by ShadwDrgn** [comment]
Something would have to be instrumented into the container to accomidate the fact that there are no world files in the world directory when the container starts because it's tmpfs/shm. Some sort of flag the tells it this is configured for volatile memory, and that it should wait until the world files are fully done restoring.

Further, unless there is some mechnism in a container to signal another container that it should run, there is no way for this container to tell whatever outside container to do a restore at all.

In my solution I've put the backup software in the same container as the server, but this doesn't seem like a good solution either. I just don't know what other way you can solve the problem of needing a restore on container startup before minecraft runs.

**Comment by shotah** [comment]
Maybe that is where the confusion is, the container should see the world files just fine as they are mounted in, and it would not know if it was running on tmpfs. Same with running something like rsync, that is not inside of the container. 

```yaml
--mount type=tmpfs,data=/data \
```

**Comment by shotah** [comment]
Also per Reddit they are suggesting with SSD's there would be little to gain from the ramdisk. 
https://www.reddit.com/r/admincraft/comments/2sv9x3/ramdisk_for_world_files/

**Comment by shotah** [comment]
Kubernetes allows for Ram backed disks. I guess I see what you are trying to do. Have the container manage its own ram-disk and Rsync. 

Kubernetes documentation for Ram backed disks
https://kubernetes.io/docs/concepts/storage/volumes/#emptydir

**Comment by ShadwDrgn** [comment]
Right, but if you run a container with that was your mount that folder will be empty. The game server would start with an empty world directory everytime. tmpfs is a ramdisk and every time the container stops those files are gone forever.

The flow would have to be:
Restore world files from backup
minecraft server starts
backups every 15-20 mins or w/e

I agree it's possible to do thee restore flow outside of the container (using shm insteead of tmpfs), however this would be wildly inconvenient imo. Also I'd prefer not to switch to kubernetes.

My nas doesn't have SSD unforunately, but if this is the recommendation, i suppose we can just close this issue as being not worth it or something. I'll continue to use ramdisk in my own container, though it would have been really  really nice to use a more professional container such as this one.

**Comment by ShadwDrgn** [comment]
If you're still interested in something like this, i t hink it could be as easy as having a flag to tell the container it's world is in volatile memory and a change to the backup container alongside this one would be needed.

If the flag is set then the entrypoint would not start the game server until "restorecomplete.txt" is present in the world directory, and the change needed to the itzg backup container would simply need to write that file after a backup is completed.

The only remaining mystery is how to get the backup container to do a restore every time the game server container starts.

**Comment by shotah** [comment]
So maybe there is confusion about what the minecraft server is doing. It's only running some start up scripts. Sure you can add a download world into that which would solve problem A. That does not solve Sync'ing unless there was a cron added, then the question is, where is it sync'ing too? Lastly I'd need to validate but I don't think we have anything in the entry point yet for Shutdown. But that would probably be the solution for sync'ing on a non-forced shutdown.

Docker Shutdown Reference: https://stackoverflow.com/questions/41451159/how-to-execute-a-script-when-i-terminate-a-docker-container

External Solution would be to just use busybox with depends on for Rsync
Busybox:
https://hub.docker.com/_/busybox

Use depends on:
https://docs.docker.com/compose/startup-order/

Same Volumes:
https://docs.docker.com/storage/volumes/

**Comment by shotah** [comment]
So if it was to be added, we'd need a tempfs volume and a back up volume. Startup script would sync files, then create a cron job to continue to sync files. Then we add a sync, shutdown command. hrrrm...

**Comment by ShadwDrgn** [comment]
Forced shutdown wouldn't cause any problems if world backups are being done properly. You could power the machine off mid world-write and it wouldn't be a problem because a restore takes place on container up if this is implemented. I don't think anything would be necessary on shutdown. I assumed itzg would want to somehow make this work with his existing backup container, but if it's going to be moving towards doing backups inside the game container, then i recommend borg. rsync is not an acceeptable backup solution in any way in my mind.

Note that startup order using native docker won't help here  docker will not have any idea whether a particular process in a container has completed before starting the next container. It will ONLY see if the container is started (even if literally no program in it has even finished it's init) before starting the next container.

The way I solved for this in my jank terribly organized container is to have backups in the same container as the game server and I pass in a BACKUPDIR env variable, and borg does the restore to the world dir from that location, and I also pass a "PREFIX" env variable to allow the backup archives to have a prefix so i can do multiple server versions.
finally to do the backups themselves i have cron running on the host that uses docker exec to run borg inside the container using the PREFIX and BACKUPDIR env variables to do an incremental and another cron job to do pruning.

**Comment by ShadwDrgn** [comment]
My dockerfile  and entrypoint.sh for reference
Dockerfile: https://pastebin.com/Bchv9x0Z
Scripts in the container:
https://pastebin.com/NeCX4UBX

do_backup is run from cron on the host once every 15 mins.

**Comment by shotah** [comment]
I see we have some scripts continuously running, that would be the pattern you'd want for your sync'ing. Then just add a start up script that ways for the sync, and a shutdown script to call the sync. I know you say you don't need it, but if you are running at 15 minute intervals, and you are asking people to hop off to reboot, they'll loose 15 minutes of build or putting things in chests, and if you are doing inventory in SQL, then you could create item duplicates or loss. Just something to think about.

Continuous run examples: https://github.com/itzg/docker-minecraft-server/blob/master/scripts/start-configuration#L35-L43

**Comment by ShadwDrgn** [comment]
I agree with you assessments about what would be needed, i would only stress that I personally would avoid archival syncing solutions like the plague because they would require users to further implement more backup solutions in addition. This may be okay if we're expected to just use the itzg backup container to do it's backups on the destination volume that's synced to, though.

I hadn't thought about external inventory stuff via sql or w/e. Didn't know that was even a thing. Yes if people don't want to lose a few minutes of progress or have syncing issues with any external dependencies then a shutdown script to do a backup/sync on shutdown would be needed too.

**Comment by shotah** [comment]
yeah its extremely common to use SQL when you have multiple worlds but you want players to be able to take their inventory with them.

mc-backup looks like its using rclone. I'm not married to any sync'ing tool. I'm just trying to stay with a Domain sort of model. If part of it is in the Minecraft container, then all of it should be in the container. If some of it is external in MC-Backup, then it should all be external in MC-backup (Haven't seen if it can do an automated restore).

**Comment by shotah** [comment]
Something like: https://www.curseforge.com/minecraft/bukkit-plugins/mysql-inventory-bridge

**Comment by toast003** [comment]
I'm doing something very similar to this by using [Anything sync daemon](https://github.com/graysky2/anything-sync-daemon) outside of the container on the folder that's mounted to /data
